checkers: ability to replace timestamp

catsout opened the pull request:
add 'timestamp' parameter to _update_version, for jsonchecker using timestamp from json rather than using the one from url response date.  

Fix for tarballs from GitHub.  

Example:
```yaml
    sources:
      - type: archive
        url: https://github.com/laurent22/joplin/archive/v2.8.4.tar.gz
        sha256: d039577ecc1e7589e7432f45d4e12308d16b5df0c1b0cc928bc451d8136761d4
        x-checker-data:
          type: json
          url: https://api.github.com/repos/laurent22/joplin/releases
          version-query: .[0] | .tag_name | sub("^v"; "")
          url-query: '"https://github.com/laurent22/joplin/archive/v" + $version +
            ".tar.gz"'
          timestamp-query: .[0] | .published_at
```

gasinvein commented on README.md:
````diff
@@ -315,6 +315,7 @@ for git type sources, specify tag query and, optionaly, commit and version queri
 ```
 
 `timestamp-query` is optional, but if provided - must return a string with timestamp in ISO format.
+`timestamp-query` is also available on non-git type sources.
````
I believe this is implied, and the current (pre-this-PR) behaviour is a flaw. So no need to state this explicitly.

gasinvein reviewed (changes requested):
Thanks for addressing this.
Please split changes to `checkers` and `jsonchecker` into different commits (changes to tests should go into the same commit that has added changes being tested).

catsout committed:
checkers: ability to replace timestamp

add 'timestamp' parameter to _update_version

gasinvein commented on tests/test_jsonchecker.py:
```diff
@@ -38,6 +38,11 @@ async def test_check(self):
                         sha256="0000000000000000000000000000000000000000000000000000000000000000"
                     ),
                 )
+            elif data.filename == "jq-1.4.tarball_url.tar.gz":
```
Should match `dest-filename` in the manifest, I suppose:
```suggestion
            elif data.filename == "jq-1.4.tarball.tar.gz":
```

gasinvein commented on tests/test_jsonchecker.py:
```diff
@@ -38,6 +38,11 @@ async def test_check(self):
                     ),
                 )
+            elif data.filename == "jq-1.4.tarball_url.tar.gz":
+                self.assertNotEqual(
```
Since we get specific release date from GH API, can we compare to it? I.e. use `assertEqual` instead of `assertNotEqual` here.

catsout committed:
jsonchecker: always using provided timestamp

Also using provided timestamp on non-git type sources, the one from url
response date will be replaced.
New test case.

gasinvein commented on tests/test_jsonchecker.py:
```diff
@@ -38,6 +38,11 @@ async def test_check(self):
+            elif data.filename == "jq-1.4.tarball.tar.gz":
+                self.assertEqual(
+                    data.new_version.timestamp,
+                    datetime.fromisoformat("2018-11-02T01:54:23+00:00"),
```
```suggestion
                    datetime.datetime.fromisoformat("2018-11-02T01:54:23+00:00"),
```

gasinvein commented on tests/io.github.stedolan.jq.yml:
```diff
@@ -10,6 +10,17 @@ modules:
+          url: https://api.github.com/repos/stedolan/jq/releases/13660432
+          version-query: '.tag_name | sub("^jq-"; "")'
+          url-query: '.tarball_url'
+          timestamp: '.published_at'
```
```suggestion
          timestamp-query: '.published_at'
```

gasinvein reviewed (approved):
LGTM, thanks.
